### ISSUE TITLE
python: pip 6.0.6 fails on OS/2

### ISSUE THREAD

**Issue by dmik** [bitwiseworks/mozilla-os2#135]
Turns out that the new `pip` package included in ESR 38 (version 6.0.6 compared to 1.5.4 in ESR 31) fails to convert a local URL (`file:///D:/path`) to just path (w/o `file:///`) by mistakenly appending `/` in front of the drive letter. As a result, installing virtualenv fails like this:
```
Creating Python environment
New python executable in D:/Coding/mozilla/master-build/_virtualenv/bin/python2.7.exe
Also creating executable in D:/Coding/mozilla/master-build/_virtualenv/bin/python.exe
Installing setuptools, pip...
  Complete output from command D:/Coding/mozilla/ma...nv/bin/python2.7.exe -c "import sys, pip; sys...d\"] + sys.argv[1:]))" setuptools pip:
  Ignoring indexes: https://pypi.python.org/simple
Collecting setuptools
  Exception:
  Traceback (most recent call last):
    File "D:/Coding/mozilla/master/python/virtualenv/virtualenv_support/pip-6.0.6-py2.py3-none-any.whl/pip/basecommand.py", line 232, in main
      status = self.run(options, args)
    File "D:/Coding/mozilla/master/python/virtualenv/virtualenv_support/pip-6.0.6-py2.py3-none-any.whl/pip/commands/install.py", line 339, in run
      requirement_set.prepare_files(finder)
    File "D:/Coding/mozilla/master/python/virtualenv/virtualenv_support/pip-6.0.6-py2.py3-none-any.whl/pip/req/req_set.py", line 355, in prepare_files
      do_download, session=self.session,
    File "D:/Coding/mozilla/master/python/virtualenv/virtualenv_support/pip-6.0.6-py2.py3-none-any.whl/pip/download.py", line 770, in unpack_url
      unpack_file_url(link, location, download_dir)
    File "D:/Coding/mozilla/master/python/virtualenv/virtualenv_support/pip-6.0.6-py2.py3-none-any.whl/pip/download.py", line 718, in unpack_file_url
      unpack_file(from_path, location, content_type, link)
    File "D:/Coding/mozilla/master/python/virtualenv/virtualenv_support/pip-6.0.6-py2.py3-none-any.whl/pip/utils/__init__.py", line 645, in unpack_file
      flatten=not filename.endswith('.whl')
    File "D:/Coding/mozilla/master/python/virtualenv/virtualenv_support/pip-6.0.6-py2.py3-none-any.whl/pip/utils/__init__.py", line 522, in unzip_file
      zipfp = open(filename, 'rb')
  IOError: [Errno 2] No such file or directory: '/D:/Coding/mozilla/master/python/virtualenv/virtualenv_support/setuptools-11.0-py2.py3-none-any.whl'

----------------------------------------
...Installing setuptools, pip...done.
Traceback (most recent call last):
  File "D:/Coding/mozilla/master/python/virtualenv/virtualenv.py", line 2353, in <module>
    main()
  File "D:/Coding/mozilla/master/python/virtualenv/virtualenv.py", line 826, in main
    symlink=options.symlink)
  File "D:/Coding/mozilla/master/python/virtualenv/virtualenv.py", line 994, in create_environment
    install_wheel(to_install, py_executable, search_dirs)
  File "D:/Coding/mozilla/master/python/virtualenv/virtualenv.py", line 962, in install_wheel
    'PIP_NO_INDEX': '1'
  File "D:/Coding/mozilla/master/python/virtualenv/virtualenv.py", line 904, in call_subprocess
    % (cmd_desc, proc.returncode))
OSError: Command D:/Coding/mozilla/ma...nv/bin/python2.7.exe -c "import sys, pip; sys...d\"] + sys.argv[1:]))" setuptools pip failed with error code 2
Traceback (most recent call last):
  File "D:/Coding/mozilla/master/python/mozbuild/mozbuild/virtualenv.py", line 475, in <module>
    manager.ensure()
  File "D:/Coding/mozilla/master/python/mozbuild/mozbuild/virtualenv.py", line 128, in ensure
    return self.build()
  File "D:/Coding/mozilla/master/python/mozbuild/mozbuild/virtualenv.py", line 371, in build
    self.create()
  File "D:/Coding/mozilla/master/python/mozbuild/mozbuild/virtualenv.py", line 147, in create
    raise Exception('Error creating virtualenv.')
Exception: Error creating virtualenv.
------ config.log ------
```

**Comment by dmik** [comment]
Looks like the guilty part is not even `pip` but the `url parse` python 2.x module. Need to debug it (and compare to Windows).

**Comment by dmik** [comment]
URL parsing seems to be a generic flaw of python, see e.g. this link https://github.com/pypa/pip/issues/783 (which is a pip defect btw). In short, `urlparse` doesn't understand DOS-like drive letters in URLs and hence leaves the leading slash in even if the path starts with a drive letter. The this flaw clearly affects Windows as well. Actually, there is a flaw in the `file:` URI scheme itself which also doesn't account for drive letters in local paths on DOS-like systems in the first place (see the description of this scheme here https://en.wikipedia.org/wiki/File_URI_scheme). This flaw created ambiguity and I recall that I already faced this problem before. A proper way (which would also make the `urlparse` problem non-existent) would be to treat the drive letter like a `netloc` part of the URL (i.e. like a pseudo-hostname which it in fact is if we take the hostname part to a more abstract level of a node name). I.e. the local file URI's should have looked like `file://D:/path/to/file.txt` instead of the current `file:///D:/path/to/file.txt`).

Anyway, I patched `pip` to overcome the above problem. I uploaded the patched zip in d0c59af14f23d71bdde8fb2a332d768f9a4da058. And here's my diff, for reference:
``` diff
--- download.py.b   2015-01-03 04:32:02.000000000 +0400
+++ download.py 2015-08-28 03:02:56.000000000 +0400
@@ -436,6 +436,13 @@
         netloc = '\\\\' + netloc

     path = urllib_request.url2pathname(netloc + path)
+
+    if not netloc and os.name in ('nt', 'os2'):
+        # the first path component may be a volume letter which must
+        # not have a leading a slash to be a valid local path
+        if path.find(':') >= 0:
+            path = path.strip('/')
+
     return path
```

With this patch, virtulaenv is now successfully created. I still get failures at the very last stage of configure related to processing `moz.build` files but that's expected (sort order, typos etc when blindly porting). Should be easy to sort out.
